### DISCUSSION TITLE
[Proposal]: Shorthand property setter

### DISCUSSION THREAD

**Discussion by murshex** [dotnet/csharplang#4481] [Language Ideas]
# Shorthand property setter

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
Easier way to quickly define a setter only property (Similar to the shorthand getter property feature) **This feature assumes we want to have a default get method**

## Motivation
I think it makes sense because we already have the shorthand getter.

## Detailed design
Its already there for the getter so why not have it for the setter.

```cs
public int hello <= 100 * value;
```

## Drawbacks
None that I can think of.

## Alternatives
Using the normal set:
```cs
private int _hello;
public int hello
{
    get => _hello;
    set => _hello = 100 * value;
}
```

## Unresolved questions
None that I can think of

## Design meetings
Sorry but I'm not sure about this one either.

**Comment by theunrepentantgeek** [comment]
What is the value of this? What problem does it solve?

```
public int hello <= 100 * value;
```

This world declare a write-only property with an inaccessible field for storage, so the only code that can read the value will be using reflection. Doesn't seem to be widely useful. 

Note that this can't be equivalent to your alternative as the backing fields for autoproperties have inexpressible names to avoid name collisions.

**Reply by jnm2** [reply]
Just like you can assign to a get-only auto property in the constructor, I'd assume you could read from a set-only auto property everywhere in the declaring class, too.

**Comment by jnm2** [comment]
Set-only properties are discouraged. They are more aptly conceptualized as methods, because when there is no way for the caller that set the value to retrieve it, the property is not conceptually doing the job of a property.

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1044:

> CA1044: Properties should not be write only

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property:

> ❌ DO NOT provide set-only properties or properties with the setter having broader accessibility than the getter.

These explain a bit more:
https://stackoverflow.com/questions/4564928/are-set-only-properties-bad-practice
https://stackoverflow.com/questions/4695551/write-only-properties-whats-the-point

A good way to discourage a pattern that is frowned upon is to keep it verbose.

**Comment by murshex** [comment]
@theunrepentantgeek @jnm2 
This is not set only/write only. 

This:
```cs
public​ ​int​ ​hello​ ​<=​ ​100​ ​*​ ​value​;
```

Is supposed to behave the same as this:
```cs
private​ ​int​ ​_hello​;
​public​ ​int​ ​hello​
{
    ​get​ ​=>​ ​_hello​;
    ​set​ ​=>​ ​_hello​ ​=​ ​100​ ​*​ ​value​;
}
```

Thank you for your reply.

**Reply by huoyaoyuan** [reply]
This is not roundtripping. Corresponding getter should be `hello / 100`.

Calculating reversion of arbitrary expression is considered impossible.

**Reply by theunrepentantgeek** [reply]
@c0nd3v - Just so you know, when you post a comment on the thread, subscribed members (such as me) get an email notification, so that we can stay up to date with comments. #offtopic

Multiple comments, even if later deleted, show up as separate emails - with different comment IDs in the links:

![image](https://user-images.githubusercontent.com/1272094/109409564-af9bb680-79f8-11eb-9e26-84b1c5d6df6a.png)

(The timestamps reflect my time-zone, which is likely not yours.)

**Reply by theunrepentantgeek** [reply]
Assuming that this
```
public​ ​int​ ​hello​ ​<=​ ​100​ ​*​ ​value​;
```
behaves as now described, I have some serious issues with the suggestion.

* Having something that looks like a **set** declaration invisibly declare a **get** as well doesn't feel like it fits with the way the rest of C# works
* Having the member `_hello` declared as a side effect runs the risk of a name collision, as well as being inconsistent with the way other auto-properties work
* It's not usual for property setters to make dramatic transformations on the value set - in the vast majority of cases, reading a value from the property should give the same value back again

Assuming I've now understood the feature request properly, I'm still wondering about my original question: What problem does this solve?

Aside from involving fewer keystrokes (which isn't in itself a sufficient goal for most C# features), what is so wrong with the longer form that we need to provide a shorter replacement? And, how is this solution sufficient to leap the [-100 points](https://docs.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points) threshold that's often discussed?

**Reply by jnm2** [reply]
@c0nd3v Sorry I missed that. "Setter only property" has a very specific meaning, and the evidence that you wanted something different from that was subtle.

**Reply by jnm2** [reply]
@theunrepentantgeek I doubt that the name `_hello` is part of the proposal. It only appears in the desugaring example and it looks like it's supposed to represent the normal autoprop backing field.

**Reply by jnm2** [reply]
@theunrepentantgeek I just realized I disagree with your third point. It's not at all strange to transform the value that was set. For example:

```cs
public string Name <= value?.Trim() ?? string.Empty;
```

**Reply by theunrepentantgeek** [reply]
@jnm2 I agree with you; that's why I used the phrase "dramatic transformation". 😁 

Apologies for not being clearer.

**Reply by murshex** [reply]
> I doubt that the name _hello is part of the proposal.

Yes, thats correct.

**Comment by jnm2** [comment]
#140 would enable this syntax:

```cs
public int Hello { get; set => field = 100 * value; }
```

This wouldn't have the downside of secretly emitting a `get` accessor that isn't present in the syntax.

**Reply by murshex** [reply]
~~While that is nice, my proposal represents the normal autoprop backing field, while this requires you to declare a class member variable (If I understand correctly)~~

Okay well, yeah that is a good alternative too. but, I still think my proposal still has value.

**Reply by murshex** [reply]
I understand why secretly emitting a get accessor might look bad but, without it, its basically a useless variable, and it seriously should not cause any problems.

**Reply by jnm2** [reply]
It seems like we've barely been able to convince the team to look at #140, and #140 generalizes so well that I'm doubtful of convincing them to add an additional syntactic form beyond that to save 20 characters on one line for a case that I see once every two years in the wild (ymmv). No harm in asking though!

**Reply by theunrepentantgeek** [reply]
I suspect that widespread developer confusion caused by magic semantics should really be considered a problem. 

FWIW don't think I'm exaggerating very much - C# has a strong history of being explicit (eg requiring **new** or **override** for **virtual** methods, requiring async if you use await, and so on). A magic **get** that appears when someone innocently refactors code would be surprising. 

#140 has all of the benefits and none of the downsides of this proposal.

**Reply by CyrusNajmabadi** [reply]
We'll be looking at #140 in two weeks :-)

**Comment by Joe4evr** [comment]
> I still think my proposal still has value.

The language team does not: https://github.com/dotnet/csharplang/issues/1278#issuecomment-360619835

**Comment by dylanrafael05** [comment]
This could work well if only applied to `set`, for example:
```cs
public string Name {
    get;
    set <= value + " ";
}
```
This would be rendered useless without #140, since its useless to get a value in a different way then setting it. If #140 is ever passed, then this syntax could be useful;
```cs
public string Name {
    get => field.Substring(0, file.Length - 1);
    set <= value + " ";
}
```
The fact that `<=` is also the less-than-or-equals operator might also complicate things, consider `<-`, `<~`, `=<`, or `<==`
